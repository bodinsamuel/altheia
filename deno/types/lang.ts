export type LangFunction = (name: string, args?: any, result?: any) => string;

export interface LangList {
  [k: string]: LangFunction;
}
