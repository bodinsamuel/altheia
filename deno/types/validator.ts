import { TypeBase } from '../validators/index.ts';

export interface ValidatorOptions {
  required?: boolean;
  unknown?: boolean;
  flatten?: boolean;
}

export interface ValidatorConfirm {
  initial: string;
  comparison: string;
}

export interface ValidatorSchema {
  [k: string]: TypeBase;
}
