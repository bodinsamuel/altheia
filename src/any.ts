import Base from './base';

/**
 * Any is just a empty class to allow chaining
 */
export class TypeAny extends Base {
  /**
   * Constructor
   */
  constructor() {
    super();
    this.name = 'any';
  }
}

const def = {
  Class: TypeAny,
};

export default def;
