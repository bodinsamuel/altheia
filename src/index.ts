import isPlainObject from 'lodash/isPlainObject';

import LangBase from './lang';
import Validator from './validator';
import TypeBase from './base';
import {
  LangList,
  AltheiaInstance,
  ValidatorSchema,
  ValidatorErrorFormatted,
  ValidatorErrorRaw,
} from './types';

// Default extractors
import AnyValidator from './any';
import ArrayValidator from './array';
import BooleanValidator from './boolean';
import DateValidator from './date';
import FunctionValidator from './function';
import InternetValidator from './internet';
import NumberValidator from './number';
import ObjectValidator from './object';
import StringValidator from './string';

export const Instance = (lang?: LangList): AltheiaInstance => {
  // @ts-ignore
  const inst: AltheiaInstance = (schema: ValidatorSchema): Validator => {
    return new Validator(schema, inst);
  };

  inst.langList = Object.assign({}, LangBase);
  inst.templates = {};
  inst.Base = TypeBase;

  inst.instance = (newLang: LangList = {}): AltheiaInstance => {
    return Instance(Object.assign({}, inst.langList, newLang));
  };

  inst.use = (plugin): AltheiaInstance => {
    const test = new plugin.Class();
    inst[test.name || test.constructor.name] = (): TypeBase => {
      return new plugin.Class(inst);
    };
    if (plugin.messages) {
      inst.lang(plugin.messages);
    }
    return inst;
  };

  inst.lang = (key, tpl): void => {
    if (isPlainObject(key)) {
      inst.langList = Object.assign({}, inst.langList, key);
    } else if (typeof key === 'string' && tpl) {
      inst.langList[key] = tpl;
    }
  };

  inst.template = (name, schema): void => {
    inst.templates[name] = schema;
  };

  inst.is = (name): TypeBase => {
    if (typeof inst.templates[name] === 'undefined') {
      throw new Error(`unknow template ${name}`);
    }

    return inst.templates[name].clone();
  };

  inst.formatError = (
    { type, args, result },
    label = 'value',
    position
  ): ValidatorErrorFormatted => {
    // Get messages from error
    let msg;
    if (typeof inst.langList[type] !== 'undefined') {
      msg = inst.langList[type](label, args, result);
    } else {
      msg = 'Invalid value';
    }

    const formatted: ValidatorErrorFormatted = {
      label,
      type,
      message: msg,
    };

    if (typeof position === 'number') {
      formatted.position = position;
    }

    // nested errors
    if (result && result.errors) {
      formatted.errors = result.errors.map((error: ValidatorErrorRaw) =>
        inst.formatError(error.test, error.label, error.position)
      );
    }
    return formatted;
  };

  // Declare basic plugins
  inst.use(AnyValidator);
  inst.use(ArrayValidator);
  inst.use(BooleanValidator);
  inst.use(DateValidator);
  inst.use(FunctionValidator);
  inst.use(InternetValidator);
  inst.use(NumberValidator);
  inst.use(ObjectValidator);
  inst.use(StringValidator);

  // Add passed lang object
  if (lang && isPlainObject(lang)) {
    inst.lang(lang);
  }

  return inst;
};

export default Instance();
