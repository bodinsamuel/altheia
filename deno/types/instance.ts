import {
  TypeBase,
  TypeAny,
  TypeArray,
  TypeBoolean,
  TypeDate,
  TypeFunc,
  TypeInternet,
  TypeNumber,
  TypeObject,
  TypeString,
} from '../validators/index.ts';
import { Validator } from '../validator.ts';

import { LangList, LangFunction } from './lang.ts';
import { ValidatorSchema } from './validator.ts';
import { ValidatorTestResult, ValidatorErrorFormatted } from './tests.ts';

// ---------- Altheia
export interface ValidatorTemplates {
  [k: string]: TypeBase;
}

export interface AltheiaInstance {
  // All plugins are here
  [k: string]: any;

  (schema: ValidatorSchema): Validator;

  langList: LangList;
  templates: ValidatorTemplates;
  Base: typeof TypeBase;

  instance: (lang?: LangList) => AltheiaInstance;
  use: (plugin: { Class: BaseConstructor; messages?: LangList }) => this;
  lang: (key: LangList | string, tpl?: LangFunction) => void;
  template: (name: string, schema: TypeBase) => void;
  is: (name: string) => TypeBase;
  formatError: (
    { type, args, result }: ValidatorTestResult,
    label?: string,
    position?: number
  ) => ValidatorErrorFormatted;

  any: () => TypeAny;
  array: () => TypeArray;
  boolean: () => TypeBoolean;
  date: () => TypeDate;
  func: () => TypeFunc;
  internet: () => TypeInternet;
  number: () => TypeNumber;
  object: () => TypeObject;
  string: () => TypeString;
}

export interface BaseConstructor {
  new (instance?: AltheiaInstance): TypeBase;
}
