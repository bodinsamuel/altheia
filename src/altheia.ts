import isPlainObject from 'lodash/isPlainObject';

import LangBase from './lang';
import { Validator } from './validator';
import { TypeBase } from './validators/base';
import { LangList } from './types/lang';
import { AltheiaInstance } from './types/instance';
import { ValidatorErrorFormatted, ValidatorErrorRaw } from './types/tests';
import { ValidatorSchema } from './types/validator';

// Default extractors
import AnyValidator from './validators/any';
import ArrayValidator from './validators/array';
import BooleanValidator from './validators/boolean';
import DateValidator from './validators/date';
import FunctionValidator from './validators/function';
import InternetValidator from './validators/internet';
import NumberValidator from './validators/number';
import ObjectValidator from './validators/object';
import StringValidator from './validators/string';

const Instance = (lang?: LangList): AltheiaInstance => {
  const inst: AltheiaInstance = ((schema: ValidatorSchema): Validator => {
    return new Validator(schema, inst);
  }) as AltheiaInstance;

  inst.langList = { ...LangBase };
  inst.templates = {};
  inst.Base = TypeBase;

  inst.instance = (newLang: LangList = {}): AltheiaInstance => {
    return Instance({ ...inst.langList, ...newLang });
  };

  inst.use = (plugin): AltheiaInstance => {
    const test = new plugin.Class();
    inst[test.name || test.constructor.name] = (): TypeBase => {
      return new plugin.Class(inst);
    };
    if (plugin.messages) {
      inst.lang(plugin.messages);
    }
    return inst;
  };

  inst.lang = (key, tpl): void => {
    if (typeof key === 'object' && isPlainObject(key)) {
      inst.langList = { ...inst.langList, ...key };
    } else if (typeof key === 'string' && tpl) {
      inst.langList[key] = tpl;
    }
  };

  inst.template = (name, schema): void => {
    inst.templates[name] = schema;
  };

  inst.is = (name): TypeBase => {
    if (typeof inst.templates[name] === 'undefined') {
      throw new Error(`unknow template ${name}`);
    }

    return inst.templates[name].clone();
  };

  inst.formatError = (
    { type, args, result },
    label = 'value',
    position
  ): ValidatorErrorFormatted => {
    // Get messages from error
    let msg;
    if (typeof inst.langList[type] !== 'undefined') {
      msg = inst.langList[type](label, args, result);
    } else {
      msg = 'Invalid value';
    }

    const formatted: ValidatorErrorFormatted = {
      label,
      type,
      message: msg,
    };

    if (typeof position === 'number') {
      formatted.position = position;
    }

    // nested errors
    if (result && result.errors) {
      formatted.errors = result.errors.map(
        (error: ValidatorErrorRaw): ValidatorErrorFormatted =>
          inst.formatError(error.test, error.label, error.position)
      );
    }
    return formatted;
  };

  // Declare basic plugins
  inst.use(AnyValidator);
  inst.use(ArrayValidator);
  inst.use(BooleanValidator);
  inst.use(DateValidator);
  inst.use(FunctionValidator);
  inst.use(InternetValidator);
  inst.use(NumberValidator);
  inst.use(ObjectValidator);
  inst.use(StringValidator);

  // Add passed lang object
  if (lang && isPlainObject(lang)) {
    inst.lang(lang);
  }

  return inst;
};

export const Altheia = Instance();
