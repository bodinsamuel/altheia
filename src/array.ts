import TypeBase from './base';
import arrayDiff from './utils/arraydiff';
import { LangList, ValidatorErrorRaw } from './types';

export const messages: LangList = {
  'array.typeof': (name) => `${name} must be a valid array`,
  'array.min': (name, args: { min: number }) =>
    `${name} must contains at least ${args.min} items`,
  'array.max': (name, args: { max: number }) =>
    `${name} must contains at most ${args.max} items`,
  'array.in': (name, args: { in: string[] }) =>
    `${name} must only contains these keys [${args.in}]`,
  'array.not': (name) => `${name} contains forbidden value`,
  'array.unique': (name) => `${name} can not contains duplicated value`,
  'array.oneOf': (name) => `${name} contains forbidden items`,
  'array.itemInvalid': (name) =>
    `${name} does not match any of the allowed types`,
};

/**
 * Array class
 */
export class TypeArray extends TypeBase {
  /**
   * Constructor
   */
  constructor() {
    super();
    this.name = 'array';
    this.typeof();
  }

  /**
   * Test to validate the type of the value
   *
   * @return {this}
   */
  typeof(): this {
    this.test('typeof', (val: any) => {
      return Array.isArray(val);
    });
    return this;
  }

  /**
   * Force an array to contains at least a number of items equal to the value passed.
   *
   * @param  {number} min
   * @return {this}
   */
  min(min: number): this {
    this.test(
      'min',
      (arr: any[]) => {
        return arr.length >= min;
      },
      { min }
    );

    return this;
  }

  /**
   * Force an array to contains at most a number of items equal to the value passed.
   *
   * @param  {number} max
   * @return {this}
   */
  max(max: number): this {
    this.test(
      'max',
      (arr: any[]) => {
        return arr.length <= max;
      },
      { max }
    );

    return this;
  }

  /**
   * Force an array to have only the values passed in the set
   *
   * @param  {...array} array
   * @return {this}
   */
  in(value: any[]): this;
  in(...value: any): this;
  in(...array: any): this {
    let only = array;
    // handle someone passing literal array instead of multiple args
    if (array.length === 1 && Array.isArray(array[0])) {
      only = (array[0] as unknown) as any[];
    }

    this.test(
      'in',
      (arr: any[]) => {
        return arrayDiff<string>(arr, only).length === 0;
      },
      { in: only }
    );

    return this;
  }

  /**
   * Force an array not to have the values passed in the set
   *
   * @param  {...array} array
   * @return {this}
   */
  not(value: any[]): this;
  not(...value: any): this;
  not(...array: any[]): this {
    let only = array;
    // handle someone passing literal array instead of multiple args
    if (array.length === 1 && Array.isArray(array[0])) {
      only = (array[0] as unknown) as any[];
    }

    this.test(
      'not',
      (arr: any[]) => {
        return arrayDiff<string>(only, arr).length === only.length;
      },
      { not: only }
    );

    return this;
  }

  /**
   * Force an array to only have each item once
   *
   * @return {this}
   */
  unique(): this {
    this.test('unique', (arr: any[]) => {
      const a = new Set(arr);
      return a.size === arr.length;
    });

    return this;
  }

  /**
   * Force all array's items to match one of the template
   *
   * @param  {...Validator} templates
   * @return {this}
   */
  oneOf(...templates: TypeBase[]): this {
    this.test(
      'oneOf',
      async (arr: any[]) => {
        const errors: ValidatorErrorRaw[] = [];

        await Promise.all(
          arr.map(async (value: any, index: number) => {
            let error: ValidatorErrorRaw | undefined = undefined;

            const label = 'item';
            for (var i = 0; i < templates.length; i++) {
              const test = await templates[i].required().validate(value);
              if (test) {
                error = { label, test, position: index };
              } else {
                // early break if one template matched (returned no error)
                return;
              }
            }

            // Help typescript understand we have error here
            if (!error) {
              return;
            }

            // if multiples templates, return a generic message
            if (templates.length > 1) {
              error = {
                label,
                test: this.createTestResult(
                  this.createTest({
                    type: 'array.itemInvalid',
                  }),
                  false
                ),
                position: index,
              };
              errors.push(error);
            } else {
              errors.push(error);
            }
          })
        );

        return {
          valid: errors.length <= 0,
          error: 'not_matching',
          errors,
        };
      },
      { templates }
    );

    return this;
  }
}

const def = {
  Class: TypeArray,
  messages,
};

export default def;
