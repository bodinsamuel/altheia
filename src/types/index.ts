import TypeBase from '../base';
import Validator from '../validator';
import { TypeString } from '../string';
import { TypeAny } from '../any';
import { TypeArray } from '../array';
import { TypeBoolean } from '../boolean';
import { TypeDate } from '../date';
import { TypeFunc } from '../function';
import { TypeInternet } from '../internet';
import { TypeNumber } from '../number';
import { TypeObject } from '../object';

// ---------- Altheia
export interface ValidatorTemplates {
  [k: string]: TypeBase;
}

export interface AltheiaInstance {
  // All plugins are here
  [k: string]: any;

  (schema: ValidatorSchema): Validator;

  langList: LangList;
  templates: ValidatorTemplates;
  Base: typeof TypeBase;

  instance: (lang?: LangList) => AltheiaInstance;
  use: (plugin: { Class: BaseConstructor; messages?: LangList }) => this;
  lang: (key: LangList | string, tpl?: LangFunction) => void;
  template: (name: string, schema: TypeBase) => void;
  is: (name: string) => TypeBase;
  formatError: (
    { type, args, result }: ValidatorTestResult,
    label?: string,
    position?: number
  ) => ValidatorErrorFormatted;

  any: () => TypeAny;
  array: () => TypeArray;
  boolean: () => TypeBoolean;
  date: () => TypeDate;
  func: () => TypeFunc;
  internet: () => TypeInternet;
  number: () => TypeNumber;
  object: () => TypeObject;
  string: () => TypeString;
}

export interface BaseConstructor {
  new (instance?: AltheiaInstance): TypeBase;
}

// ---------- Lang
export type LangFunction = (name: string, args?: any, result?: any) => string;

export interface LangList {
  [k: string]: LangFunction;
}

// ---------- Validator
export interface ValidatorOptions {
  required?: boolean;
  unknown?: boolean;
  flatten?: boolean;
}

export interface ValidatorConfirm {
  initial: string;
  comparison: string;
}

export interface ValidatorSchema {
  [k: string]: TypeBase;
}

// ---------- Tests Public

export interface ValidatorTestResult {
  valid: boolean;
  type: string;
  func: TestFunction;
  args: any;
  result?: ValidatorInternalTestResult;
}

export interface ValidatorErrorRaw {
  test: ValidatorTestResult;
  label: string;
  position?: number;
}

export interface ValidatorErrorFormatted {
  label: string;
  type: string;
  message: string;
  position?: number;
  errors?: ValidatorErrorFormatted[];
}

export type TestFunction = (
  params: any
) =>
  | Promise<boolean | ValidatorInternalTestResult>
  | boolean
  | ValidatorInternalTestResult;

// ---------- Tests Private
export interface ValidatorInternalTest {
  type: string;
  func: TestFunction;
  args: any;
}

export interface ValidatorInternalTestResult {
  valid: boolean;
  error: string;
  overrideWith?: ValidatorTestResult;
  errors?: ValidatorErrorRaw[];
}
