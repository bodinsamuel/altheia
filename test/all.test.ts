import Alt from './../src';
import { ValidatorErrorFormatted } from '../src/types';

describe('All', () => {
  test('should pass', async () => {
    const hasError = await Alt({
      foo: Alt.string(),
      number: Alt.number(),
      bool: Alt.boolean(),
      array: Alt.array(),
      function: Alt.function(),
      date: Alt.date(),
      internetUrl: Alt.internet().url(),
      object: Alt.object(),
      arrayWithTemplate: Alt.array().oneOf(Alt.string()),
      arrayWithTemplates: Alt.array().oneOf(Alt.string(), Alt.number()),
    })
      .body({
        foo: 'bar',
        number: 1,
        bool: true,
        array: [],
        function: () => true,
        date: new Date(),
        internetUrl: 'http://localhost/',
        object: { foo: 'bar' },
        arrayWithTemplate: ['foo'],
        arrayWithTemplates: ['foo', 1],
      })
      .validate();
    expect(hasError).toBe(false);
  });

  test('should fail', async () => {
    const hasError = await Alt({
      foo: Alt.string(),
      number: Alt.number(),
      bool: Alt.boolean(),
      array: Alt.array(),
      function: Alt.function(),
      date: Alt.date(),
      internetUrl: Alt.internet().url(),
      object: Alt.object(),
      arrayWithTemplate: Alt.array().oneOf(Alt.string()),
      arrayWithTemplates: Alt.array().oneOf(Alt.string(), Alt.number()),
    })
      .body({
        foo: 1,
        number: 'bar',
        bool: [],
        array: 'foo',
        function: new Date(),
        date: () => true,
        internetUrl: 1,
        object: [],
        arrayWithTemplate: [1],
        arrayWithTemplates: [true, {}],
      })
      .validate();
    expect(hasError).toHaveLength(10);
    expect(hasError).toMatchSnapshot();
  });

  test('should fail nested', async () => {
    const hasError = await Alt({
      foo: Alt.array().oneOf(Alt.string()),
      object: Alt.object().schema(
        Alt({
          bar: Alt.string(),
        })
      ),
      nested: Alt.array().oneOf(
        Alt.object().schema(
          Alt({
            foo: Alt.string(),
            nested2: Alt.array().oneOf(
              Alt.object().schema(
                Alt({ hello: Alt.string().in('nice') }).options({
                  required: true,
                })
              ),
              Alt.object().schema(
                Alt({ bonjour: Alt.string().in('cool') }).options({
                  required: true,
                })
              )
            ),
          }).options({ required: true })
        )
      ),
    })
      .body({
        foo: [1, 'foo', 'b', [], 'ar'],
        object: { bar: 1 },
        nested: [
          {
            foo: 'bar',
            nested2: [{ hello: 'nice' }, { hello: 2 }, { bonjour: 2 }],
          },
          {
            foo: 'bar',
            nested2: [{ hello: 'nice' }],
          },
        ],
      })
      .validate();
    expect(hasError).toHaveLength(3);
    expect(hasError).toMatchSnapshot();

    expect((hasError as ValidatorErrorFormatted[])[2].label).toBe('nested');
    expect((hasError as ValidatorErrorFormatted[])[2].errors[0].label).toBe(
      'item'
    );
    expect(
      (hasError as ValidatorErrorFormatted[])[2].errors[0].errors[0].label
    ).toBe('nested2');
    expect(
      (hasError as ValidatorErrorFormatted[])[2].errors[0].errors[0].errors[0]
        .label
    ).toBe('item');
    expect(
      (hasError as ValidatorErrorFormatted[])[2].errors[0].errors[0].errors[0]
        .position
    ).toBe(1);
  });
});
