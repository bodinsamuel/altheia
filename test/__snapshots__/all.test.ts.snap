// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All should fail 1`] = `
Array [
  Object {
    "label": "foo",
    "message": "foo must be a valid string",
    "type": "string.typeof",
  },
  Object {
    "label": "number",
    "message": "number must be a valid number",
    "type": "number.typeof",
  },
  Object {
    "label": "bool",
    "message": "bool must be a valid boolean",
    "type": "boolean.typeof",
  },
  Object {
    "label": "array",
    "message": "array must be a valid array",
    "type": "array.typeof",
  },
  Object {
    "label": "function",
    "message": "function must be a valid function",
    "type": "function.typeof",
  },
  Object {
    "label": "date",
    "message": "date must be a valid date",
    "type": "date.typeof",
  },
  Object {
    "label": "internetUrl",
    "message": "internetUrl must be a valid string",
    "type": "internet.typeof",
  },
  Object {
    "label": "object",
    "message": "object must be a valid object",
    "type": "object.typeof",
  },
  Object {
    "errors": Array [
      Object {
        "label": "item",
        "message": "item must be a valid string",
        "position": 0,
        "type": "string.typeof",
      },
    ],
    "label": "arrayWithTemplate",
    "message": "arrayWithTemplate contains forbidden items",
    "type": "array.oneOf",
  },
  Object {
    "errors": Array [
      Object {
        "label": "item",
        "message": "item does not match any of the allowed types",
        "position": 0,
        "type": "array.itemInvalid",
      },
      Object {
        "label": "item",
        "message": "item does not match any of the allowed types",
        "position": 1,
        "type": "array.itemInvalid",
      },
    ],
    "label": "arrayWithTemplates",
    "message": "arrayWithTemplates contains forbidden items",
    "type": "array.oneOf",
  },
]
`;

exports[`All should fail nested 1`] = `
Array [
  Object {
    "errors": Array [
      Object {
        "label": "item",
        "message": "item must be a valid string",
        "position": 0,
        "type": "string.typeof",
      },
      Object {
        "label": "item",
        "message": "item must be a valid string",
        "position": 3,
        "type": "string.typeof",
      },
    ],
    "label": "foo",
    "message": "foo contains forbidden items",
    "type": "array.oneOf",
  },
  Object {
    "errors": Array [
      Object {
        "label": "bar",
        "message": "bar must be a valid string",
        "type": "string.typeof",
      },
    ],
    "label": "object",
    "message": "object does not match its schema",
    "type": "object.schema",
  },
  Object {
    "errors": Array [
      Object {
        "errors": Array [
          Object {
            "errors": Array [
              Object {
                "label": "item",
                "message": "item does not match any of the allowed types",
                "position": 1,
                "type": "array.itemInvalid",
              },
              Object {
                "label": "item",
                "message": "item does not match any of the allowed types",
                "position": 2,
                "type": "array.itemInvalid",
              },
            ],
            "label": "nested2",
            "message": "nested2 contains forbidden items",
            "type": "array.oneOf",
          },
        ],
        "label": "item",
        "message": "item does not match its schema",
        "position": 0,
        "type": "object.schema",
      },
    ],
    "label": "nested",
    "message": "nested contains forbidden items",
    "type": "array.oneOf",
  },
]
`;
